File: Lab1_java_v3/src/Classes/Book.java
package Classes;

import java.util.ArrayList;
import java.util.List;

/**
 * Класс "Книга": название, автор, год, теги
 */
public class Book {

    private String name;        //Название
    private String author;      //автор
    private String year;           //год
    private List<String> tags;  //теги: жанр, особенности (греческая мифология, славянский фольклор, вов и пр.)

    public Book(String name, String author, String year, List<String> tags) {
        this.name = name;
        this.author = author;
        this.year = year;
        this.tags = tags;
    }
    public Book(Book book) {
        this.name = book.name;
        this.author = book.author;
        this.year = book.year;
        this.tags = new ArrayList<>(book.tags);
    }
    public Book() {
        this.name = "";
        this.author = "";
        this.year = "";
        this.tags = null;
    }

    //region getters & setters
    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getAuthor() {
        return author;
    }

    public void setAuthor(String author) {
        this.author = author;
    }

    public String getYear() {
        return year;
    }

    public void setYear(String year) {
        this.year = year;
    }

    public List<String> getTags() {
        return tags;
    }

    public void setTags(List<String> tags) {
        this.tags = tags;
    }
    //endregion

    //метод определения и возврата степени "схожести" книги по имеющимся в стеке тегам от 0 до 1
    public float getSimilarDegree(List<String> stack){
        float degree = 0;


        //получаем количество совпадений и вычисляем коэффициент
        List<String> similars = new ArrayList<>(tags);
        similars.retainAll(stack);
        degree = (float)similars.size() / (float)tags.size();

        return  degree;
    }
    //поиск наиболее подходящей книги
    public static Book findBook(List<String> stack, List<Book> books){
        float max = 0;
        Book book = null;
        for(int i = 0; i < books.size(); i++){
            if(books.get(i).getSimilarDegree(stack) > max){
                max = books.get(i).getSimilarDegree(stack);
                book = books.get(i);
            }
        }
        /*for(Book b: books) {
           if(b.getSimilarDegree(stack) > max){
               max = b.getSimilarDegree(stack);
               book = b;
           }
        }*/

        return  book;
    }

    public static List<String> toStringArray(List<Book> books){
        List<String> bookList = new ArrayList<String>();
        for (Book b: books) {
            bookList.add(b.getName()+"\n\t"+b.getAuthor()+"\t\n\t"+b.getYear());
        }
        return bookList;
    }
    public boolean isContained(List<Book> books){
        boolean flag = false;

        for (Book b: books) {
            if (b.getSimilarDegree(this.tags) == 1 &&
                b.getName().equals(this.name) &&
                b.getAuthor().equals(this.author) &&
                b.getYear().equals(this.year)
            ) {
                flag = true;
                break;
            }
        }

        return flag;
    }
}

File: Lab1_WFApp/Lab1_WFApp/Classes/Book.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Lab1_WFApp
{
    class Book
    {
        private string name;        //Название
        private string author;      //автор
        private string year;           //год
        private List<string> tags;  //теги: жанр, особенности (греческая мифология, славянский фольклор, вов и пр.)

        public string Name { get => name; set => name = value; }
        public string Author { get => author; set => author = value; }
        public string Year { get => year; set => year = value; }
        public List<string> Tags { get => tags; set => tags = value; }

        public Book(string name, string author, string year, List<string> tags)
        {
            this.Name = name;
            this.Author = author;
            this.Year = year;
            this.Tags = tags;
        }
        public Book(Book book)
        {
            this.Name = book.Name;
            this.Author = book.Author;
            this.Year = book.Year;
            this.Tags = new List<string>(book.Tags);
        }
        public Book()
        {
            this.Name = "";
            this.Author = "";
            this.Year = "";
            this.Tags = null;
        }

      

        //метод определения и возврата степени "схожести" книги по имеющимся в стеке тегам от 0 до 1
        public float getSimilarDegree(List<string> stack)
        {
            float degree = 0;


            //получаем количество совпадений и вычисляем коэффициент
            List<string> similars = new List<string>(Tags);
            var sim = from t in Tags
                      where stack.Contains(t)
                      select t;
            degree = (float)sim.Count()/ (float)Tags.Count;

            return degree;
        }
        //поиск наиболее подходящей книги
        public static Book findBook(List<string> stack, List<Book> books)
        {
            float max = 0;
            Book book = null;
            for (int i = 0; i < books.Count; i++)
            {
                if (books[i].getSimilarDegree(stack) > max)
                {
                    max = books[i].getSimilarDegree(stack);
                    book = books[i];
                }
            }
            /*for(Book b: books) {
               if(b.getSimilarDegree(stack) > max){
                   max = b.getSimilarDegree(stack);
                   book = b;
               }
            }*/

            return book;
        }

        public static List<string> tostringArray(List<Book> books)
        {
            List<string> bookList = new List<string>();
            foreach (Book b in books)
            {
                bookList.Add(b.Name + "\n\t" + b.Author + "\t\n\t" + b.Year);
            }
            return bookList;
        }
        public bool isContained(List<Book> books)
        {
            bool flag = false;

            foreach (Book b in books)
            {
                if (b.getSimilarDegree(this.Tags) == 1 &&
                    b.Name.Equals(this.Name) &&
                    b.Author.Equals(this.Author) &&
                    b.Year.Equals(this.Year)
                )
                {
                    flag = true;
                    break;
                }
            }

            return flag;
        }
    }
}
